premium/usage.go

```diff
@@ -0,0 +1,264 @@
+	}
+
+	if u.isClosed {
+		return fmt.Errorf("usage updater is closed")
+	}
+
+	u.rowsToUpdate.Add(rows)
+
+	// Trigger an update unless an update is already in process
+	select {
+	case u.triggerUpdate <- struct{}{}:
+	default:
+		return nil
+	}
+
+	return nil
+}
+
+func (u *BatchUpdater) HasQuota(ctx context.Context) (bool, error) {
+	usage, err := u.apiClient.GetTeamPluginUsageWithResponse(ctx, u.teamName, u.pluginTeam, cqapi.PluginKind(u.pluginKind), u.pluginName)
+	if err != nil {
+		return false, fmt.Errorf("failed to get usage: %w", err)
+	}
+	if usage.StatusCode() != http.StatusOK {
+		return false, fmt.Errorf("failed to get usage: %s", usage.Status())
+	}
+	return *usage.JSON200.RemainingRows > 0, nil
+}
+
+func (u *BatchUpdater) Close(_ context.Context) error {
+	u.isClosed = true
+
+	close(u.done)
+
+	return <-u.closeError
+}
+
+func (u *BatchUpdater) backgroundUpdater(ctx context.Context) {
+	started := make(chan struct{})
+
+	flushDuration := time.NewTicker(u.flushDuration)
+
+	go func() {
+		started <- struct{}{}
+		for {
+			select {
+			case <-u.triggerUpdate:
+				if time.Since(u.lastUpdateTime) < u.minimumUpdateDuration {
+					// Not enough time since last update
+					continue
+				}
+
+				rowsToUpdate := u.rowsToUpdate.Load()
+				if rowsToUpdate < u.batchLimit {
+					// Not enough rows to update
+					continue
+				}
+				if err := u.updateUsageWithRetryAndBackoff(ctx, rowsToUpdate); err != nil {
+					log.Warn().Err(err).Msg("failed to update usage")
+					continue
+				}
+				u.rowsToUpdate.Add(-rowsToUpdate)
+			case <-flushDuration.C:
+				if time.Since(u.lastUpdateTime) < u.minimumUpdateDuration {
+					// Not enough time since last update
+					continue
+				}
+				rowsToUpdate := u.rowsToUpdate.Load()
+				if rowsToUpdate == 0 {
+					continue
+				}
+				if err := u.updateUsageWithRetryAndBackoff(ctx, rowsToUpdate); err != nil {
+					log.Warn().Err(err).Msg("failed to update usage")
+					continue
+				}
+				u.rowsToUpdate.Add(-rowsToUpdate)
+			case <-u.done:
+				remainingRows := u.rowsToUpdate.Load()
+				if remainingRows != 0 {
+					if err := u.updateUsageWithRetryAndBackoff(ctx, remainingRows); err != nil {
+						u.closeError <- err
+						return
+					}
+					u.rowsToUpdate.Add(-remainingRows)
+				}
+				u.closeError <- nil
+				return
+			}
+		}
+	}()
+	<-started
+}
+
+func (u *BatchUpdater) updateUsageWithRetryAndBackoff(ctx context.Context, numberToUpdate uint32) error {
+	for retry := 0; retry < u.maxRetries; retry++ {
+		queryStartTime := time.Now()
+
+		resp, err := u.apiClient.IncreaseTeamPluginUsageWithResponse(ctx, u.teamName, cqapi.IncreaseTeamPluginUsageJSONRequestBody{
+			RequestId:  uuid.New(),
+			PluginTeam: u.pluginTeam,
+			PluginKind: cqapi.PluginKind(u.pluginKind),
+			PluginName: u.pluginName,
+			Rows:       int(numberToUpdate),
+		})
+		if err != nil {
+			return fmt.Errorf("failed to update usage: %w", err)
+		}
+		if resp.StatusCode() == http.StatusOK {
```

Comment:
I think this will be `201` or `204` going forward (I'm changing it in an upcoming PR). Maybe just accept any 2xx code to be safe.